enhancements/microshift/multus-cni-for-microshift.md

```diff
@@ -0,0 +1,445 @@
+Exact values depend on the runtime environment as configs may require name of the specific host interface.
+For these reasons, we can only provide examples and guidance on crafting CNI configs, therefore we
+cannot package a default configuration that could work on every host in `microshift-networking-multus` RPM.
+
+Below is example of NAD to use the `bridge` CNI with host bridge named `test-bridge`:
+```yaml
+apiVersion: "k8s.cni.cncf.io/v1"
+kind: NetworkAttachmentDefinition
+metadata:
+  name: bridge-conf
+spec:         
+  config: '{                                
+      "cniVersion": "0.3.0",
+      "type": "bridge",
+      "bridge": "test-bridge",
+      "mode": "bridge",                                                   
+      "ipam": {  
+        "type": "host-local",
+        "subnet": "192.168.20.0/24",
+        "rangeStart": "192.168.20.200",
+        "rangeEnd": "192.168.20.216",
+        "routes": [
+          { "dst": "0.0.0.0/0" }
+        ],
+        "gateway": "192.168.20.1"
+      }
+    }'
+```
+
+There are two Pod annotation that can be used to instruct Multus on how the networks should be set up.
+Main one is `k8s.v1.cni.cncf.io/networks` which specifies which NAD should be added to the Pod.
+Multiple NADs can be specified by separating them with comma. NADs can be even reused.
+Examples from Multus docs:
+```yaml
+apiVersion: v1                  
+kind: Pod                        
+metadata:
+  name: samplepod  
+  annotations:
+    k8s.v1.cni.cncf.io/networks: bridge-conf
+---
+apiVersion: v1
+kind: Pod
+metadata:
+  name: samplepod
+  annotations:
+    k8s.v1.cni.cncf.io/networks: bridge-conf,bridge-conf
+```
+
+Other Pod annotation is `v1.multus-cni.io/default-network` which overrides default CNI from Multus'
+configuration file (created on DaemonSet start). This annotation is optional and its absence means
+Pod's default network will be ovn-kubernetes.
+
+Information about the Pod's interfaces are reported back by Multus also as an annotation `k8s.v1.cni.cncf.io/network-status`.
+Example below shows annotations of a Pod with two `bridge` CNI interfaces:
+```yaml
+Annotations:      k8s.v1.cni.cncf.io/network-status:
+                    [{
+                        "name": "ovn-kubernetes",
+                        "interface": "eth0",
+                        "ips": [
+                            "10.42.0.14"
+                        ],
+                        "mac": "0a:58:0a:2a:00:0e",
+                        "default": true,
+                        "dns": {}
+                    },{
+                        "name": "default/bridge-conf",
+                        "interface": "net1",
+                        "ips": [
+                            "192.168.20.202"
+                        ],
+                        "mac": "b2:6a:5c:ba:34:9a",
+                        "dns": {},
+                        "gateway": [
+                            "\u003cnil\u003e"
```

Comment:
TBH I don't know. It looks funky to me but this is what I got from the annotations when using Multus...

Reply:
@s1061123  this sounds like a bug?

Reply:
`\u003c` is `<` and `\u003e` is `>` so it looks like something is turning `<nil>` into a unicode byte string when it renders the annotation as JSON.

Reply:
So this is happen because host-local IPAM plugin is used with wrong configuration. As described https://www.cni.dev/plugins/current/ipam/host-local/, `gateway` configuration should be in `range` section.

With following net-attach-def, 

```
---
apiVersion: "k8s.cni.cncf.io/v1"
kind: NetworkAttachmentDefinition
metadata:
  name: bridge-conf-1
spec:
  config: '{
            "cniVersion": "0.3.1",
            "type": "bridge",
            "bridge": "testbr1",
            "ipam": {
                "type": "host-local",
                "ranges": [
                    [
                        {
                            "subnet": "10.10.0.0/16",
                            "rangeStart": "10.10.1.20",
                            "rangeEnd": "10.10.3.50",
                            "gateway": "10.10.0.254"
                        }
                    ]
                ],
                "dataDir": "/var/lib/cni/testbr1"
            }
        }'
```

Then get the following result (gateway is configured in eth0 hence no gateway in net1)

```
    annotations:
      k8s.v1.cni.cncf.io/network-status: |-
        [{
            "name": "cbr0",
            "interface": "eth0",
            "ips": [
                "10.244.1.2"
            ],
            "mac": "5a:a3:ef:74:d1:36",
            "default": true,
            "dns": {},
            "gateway": [
                "10.244.1.1"
            ]
        },{
            "name": "default/bridge-conf-1",
            "interface": "net1",
            "ips": [
                "10.10.1.20"
            ],
            "mac": "7e:07:16:b2:ef:c5",
            "dns": {}
        }]
```

Reply:
Thanks! I used the snippet to update the enhancement